File: bench/python/bench_regression.py
#!/usr/bin/env python3
"""Regression benchmarks (B1) — Python side.

Runs the Python equivalents of every benchmark in
``bench/haskell/BenchRegression.hs`` over the shared CSVs in
``bench/data/``. Writes ``bench/results/python/regression.csv`` with the
unified BenchRow schema.

Run with::

    OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 \
        bench/venv/bin/python bench/python/bench_regression.py
"""

from __future__ import annotations

import csv
import os
import time
from dataclasses import dataclass
from pathlib import Path

import numpy as np
import pandas as pd
from sklearn.linear_model import (
    LinearRegression,
    LogisticRegression,
    PoissonRegressor,
    Ridge,
    Lasso,
    ElasticNet,
)
import statsmodels.api as sm
import statsmodels.formula.api as smf


REPO = Path(__file__).resolve().parents[2]
DATA = REPO / "bench" / "data"
OUT = REPO / "bench" / "results" / "python"
OUT.mkdir(parents=True, exist_ok=True)


@dataclass
class Row:
    name: str
    time_ms: float
    acc_main: float
    acc_aux: float
    extra: str


def median_time(fn, n_iter: int) -> tuple[float, object]:
    """Run *fn* `n_iter` times, return (median ms, last result)."""
    times = []
    last = None
    for _ in range(n_iter):
        t0 = time.perf_counter()
        last = fn()
        t1 = time.perf_counter()
        times.append(1000.0 * (t1 - t0))
    times.sort()
    return times[len(times) // 2], last


def load_xy(path: Path) -> tuple[np.ndarray, np.ndarray]:
    df = pd.read_csv(path)
    return df.iloc[:, :-1].to_numpy(), df.iloc[:, -1].to_numpy()


def load_xyg(path: Path) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    df = pd.read_csv(path)
    g = df["group"].to_numpy()
    y = df["y"].to_numpy()
    x = df.drop(columns=["group", "y"]).to_numpy()
    return x, g, y


def r2(y, yhat):
    mu = y.mean()
    sst = float(((y - mu) ** 2).sum())
    sse = float(((y - yhat) ** 2).sum())
    return 0.0 if sst == 0 else 1 - sse / sst


def rmse(y, yhat):
    return float(np.sqrt(((y - yhat) ** 2).mean()))


# ---------------------------------------------------------------------------
# LM (OLS)
# ---------------------------------------------------------------------------

def bench_lm(path: Path, name: str) -> Row:
    x, y = load_xy(path)

    def run():
        m = LinearRegression(fit_intercept=True).fit(x, y)
        # ensure prediction also forces internal computation
        return m, m.predict(x[:1])

    ms, (m, _) = median_time(run, 5)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               "sklearn.LinearRegression")


# ---------------------------------------------------------------------------
# GLM Logistic / Poisson
# ---------------------------------------------------------------------------

def bench_logit(path: Path, name: str) -> Row:
    x, y = load_xy(path)

    def run():
        m = LogisticRegression(
            penalty=None, fit_intercept=True, solver="lbfgs",
            max_iter=200, tol=1e-6,
        ).fit(x, y)
        return m

    ms, m = median_time(run, 3)
    p = m.predict_proba(x)[:, 1]
    return Row(name, ms, r2(y, p), rmse(y, p),
               "sklearn.LogisticRegression (lbfgs)")


def bench_poisson(path: Path, name: str) -> Row:
    x, y = load_xy(path)

    def run():
        m = PoissonRegressor(
            alpha=0.0, fit_intercept=True, solver="lbfgs",
            max_iter=200, tol=1e-6,
        ).fit(x, y)
        return m

    ms, m = median_time(run, 3)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               "sklearn.PoissonRegressor")


# ---------------------------------------------------------------------------
# LME (Gaussian, statsmodels MixedLM)
# ---------------------------------------------------------------------------

def bench_lme(path: Path, name: str) -> Row:
    x, g, y = load_xyg(path)
    # statsmodels MixedLM expects a design including intercept
    cols = [f"x{i}" for i in range(x.shape[1])]
    df = pd.DataFrame(x, columns=cols)
    df["group"] = g
    df["y"] = y
    formula = "y ~ " + " + ".join(cols)

    def run():
        return smf.mixedlm(formula, df, groups=df["group"]).fit(method="lbfgs",
                                                                disp=False)

    ms, fit = median_time(run, 3)
    yhat = fit.fittedvalues.to_numpy()
    icc = float(fit.cov_re.iloc[0, 0]
                / (fit.cov_re.iloc[0, 0] + float(fit.scale)))
    return Row(name, ms, r2(y, yhat), icc, "statsmodels.MixedLM (lbfgs)")


# ---------------------------------------------------------------------------
# Ridge / Lasso / ElasticNet
# ---------------------------------------------------------------------------

def bench_ridge(path: Path, name: str, lam: float) -> Row:
    x, y = load_xy(path)

    def run():
        m = Ridge(alpha=lam, fit_intercept=False, solver="cholesky").fit(x, y)
        return m

    ms, m = median_time(run, 5)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               f"sklearn.Ridge alpha={lam}")


def bench_lasso(path: Path, name: str, lam: float) -> Row:
    x, y = load_xy(path)

    def run():
        m = Lasso(alpha=lam, fit_intercept=False, max_iter=200,
                  tol=1e-4).fit(x, y)
        return m

    ms, m = median_time(run, 3)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               f"sklearn.Lasso alpha={lam}")


def bench_en(path: Path, name: str, lam1: float, lam2: float) -> Row:
    x, y = load_xy(path)
    alpha = lam1 + lam2
    l1_ratio = 0.5 if alpha == 0 else lam1 / alpha

    def run():
        m = ElasticNet(alpha=alpha, l1_ratio=l1_ratio,
                       fit_intercept=False, max_iter=200,
                       tol=1e-4).fit(x, y)
        return m

    ms, m = median_time(run, 3)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               f"sklearn.ElasticNet alpha={alpha} l1_ratio={l1_ratio}")


# ---------------------------------------------------------------------------

def main() -> int:
    rows: list[Row] = [
        bench_lm(DATA / "lm_n1000_p5.csv", "LM_n1000_p5"),
        bench_lm(DATA / "lm_n10000_p50.csv", "LM_n10000_p50"),
        bench_lm(DATA / "lm_n100000_p100.csv", "LM_n100000_p100"),

        bench_logit(DATA / "logistic_n2000_p10.csv",  "GLM_logit_n2000_p10"),
        bench_logit(DATA / "logistic_n10000_p20.csv", "GLM_logit_n10000_p20"),
        bench_poisson(DATA / "poisson_n2000_p10.csv",  "GLM_poisson_n2000_p10"),
        bench_poisson(DATA / "poisson_n10000_p20.csv", "GLM_poisson_n10000_p20"),

        bench_lme(DATA / "glmm_n2000_p5_g20.csv",   "LME_n2000_p5_g20"),
        bench_lme(DATA / "glmm_n10000_p10_g50.csv", "LME_n10000_p10_g50"),

        bench_ridge(DATA / "lm_n1000_p5.csv",   "Ridge_n1000_p5",   1.0),
        bench_ridge(DATA / "lm_n10000_p50.csv", "Ridge_n10000_p50", 1.0),

        bench_lasso(DATA / "lm_n1000_p5.csv",   "Lasso_n1000_p5",   0.05),
        bench_lasso(DATA / "lm_n10000_p50.csv", "Lasso_n10000_p50", 0.05),

        bench_en(DATA / "lm_n1000_p5.csv",   "EN_n1000_p5",   0.05, 0.05),
        bench_en(DATA / "lm_n10000_p50.csv", "EN_n10000_p50", 0.05, 0.05),
    ]

    out_path = OUT / "regression.csv"
    with open(out_path, "w", newline="") as fh:
        w = csv.writer(fh)
        w.writerow(["system", "suite", "name",
                    "time_ms", "acc_main", "acc_aux", "extra"])
        for r in rows:
            w.writerow(["python", "regression", r.name,
                        f"{r.time_ms:.6g}", f"{r.acc_main:.6g}",
                        f"{r.acc_aux:.6g}", r.extra])
    print(f"wrote {len(rows)} rows -> {out_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bench/haskell/BenchRegression.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-full-laziness -fno-cse #-}
-- | Regression benchmarks (B1).
--
-- LM, Logistic GLM, Poisson GLM, Gaussian LME (GLMM), Ridge, Lasso,
-- ElasticNet on the shared @bench/data/*.csv@ files. Outputs the unified
-- BenchRow CSV at @bench/results/haskell/regression.csv@.

module Main where

import qualified Data.Vector             as V
import qualified Data.Text               as T
import qualified Numeric.LinearAlgebra   as LA

import           Hanalyze.Model.Core              (FitResult (..))
import           Hanalyze.Model.LM                (fitLMVec)
import           Hanalyze.Model.GLM               (Family (..), LinkFn (..), fitGLMFull)
import qualified Hanalyze.Model.GLMM              as GLMM
import qualified Hanalyze.Model.Regularized       as Reg

import           BenchUtil

-- ---------------------------------------------------------------------------
-- Memoization-defeating wrappers.
--
-- GHC will common-subexpression-eliminate @fitLMVec xWith1 y@ across
-- iterations because the function is pure and the inputs are bound once
-- outside the loop. Each wrapper takes the iteration index as a phantom
-- argument and is marked NOINLINE so the optimizer cannot see through it.
-- ---------------------------------------------------------------------------

{-# NOINLINE fitLMVecPhantom #-}
fitLMVecPhantom :: Int -> LA.Matrix Double -> LA.Vector Double -> FitResult
fitLMVecPhantom _ x y = fitLMVec x y

{-# NOINLINE fitGLMFullPhantom #-}
fitGLMFullPhantom :: Int -> Family -> LinkFn
                  -> LA.Matrix Double -> LA.Vector Double -> FitResult
fitGLMFullPhantom _ fam link x y = fst (fitGLMFull fam link x y)

{-# NOINLINE fitLMEPhantom #-}
fitLMEPhantom :: Int -> LA.Matrix Double -> LA.Vector Double
              -> V.Vector Int -> V.Vector T.Text -> V.Vector Int
              -> GLMM.GLMMResult
fitLMEPhantom _ x y idx labels sizes = GLMM.fitLME x y idx labels sizes

{-# NOINLINE fitRegPhantom #-}
fitRegPhantom :: Int -> Reg.Penalty
              -> LA.Matrix Double -> LA.Vector Double -> Reg.RegFit
-- Match the Python-side bench's @max_iter=200, tol=1e-4@. The previous
-- run used hanalyze's hardcoded @1000 / 1e-7@ which made tol 1000×
-- stricter than sklearn's bench setting and gave an unfair speed
-- comparison.
fitRegPhantom _ pen x y = Reg.fitRegularizedWith 200 1e-4 pen x y

main :: IO ()
main = do
  rows <- mconcat <$> sequence
    [ benchLM "bench/data/lm_n1000_p5.csv"      "LM_n1000_p5"
    , benchLM "bench/data/lm_n10000_p50.csv"    "LM_n10000_p50"
    , benchLM "bench/data/lm_n100000_p100.csv"  "LM_n100000_p100"
    , benchLogistic "bench/data/logistic_n2000_p10.csv"  "GLM_logit_n2000_p10"
    , benchLogistic "bench/data/logistic_n10000_p20.csv" "GLM_logit_n10000_p20"
    , benchPoisson  "bench/data/poisson_n2000_p10.csv"   "GLM_poisson_n2000_p10"
    , benchPoisson  "bench/data/poisson_n10000_p20.csv"  "GLM_poisson_n10000_p20"
    , benchLME "bench/data/glmm_n2000_p5_g20.csv"    "LME_n2000_p5_g20"
    , benchLME "bench/data/glmm_n10000_p10_g50.csv"  "LME_n10000_p10_g50"
    , benchRidge "bench/data/lm_n1000_p5.csv"     "Ridge_n1000_p5"      1.0
    , benchRidge "bench/data/lm_n10000_p50.csv"   "Ridge_n10000_p50"    1.0
    , benchLasso "bench/data/lm_n1000_p5.csv"     "Lasso_n1000_p5"      0.05
    , benchLasso "bench/data/lm_n10000_p50.csv"   "Lasso_n10000_p50"    0.05
    , benchEN    "bench/data/lm_n1000_p5.csv"     "EN_n1000_p5"         0.05 0.05
    , benchEN    "bench/data/lm_n10000_p50.csv"   "EN_n10000_p50"       0.05 0.05
    ]
  writeRows "bench/results/haskell/regression.csv" rows
  putStrLn $ "wrote " ++ show (length rows)
          ++ " rows → bench/results/haskell/regression.csv"

-- ---------------------------------------------------------------------------
-- LM (OLS)
-- ---------------------------------------------------------------------------

benchLM :: FilePath -> String -> IO [BenchRow]
benchLM path name = do
  (x, y) <- readCsvXY path
  let xWith1 = LA.fromBlocks [[ LA.konst 1 (LA.rows x, 1), x ]]
  (ms, fr) <- timeitTastyIO forceFR
                (\i -> return $! fitLMVecPhantom i xWith1 y)
  let yhat = LA.flatten (fitted fr LA.¿ [0])
      r2   = computeR2 y yhat
      rmse = sqrt (LA.sumElements ((y - yhat) ** 2) / fromIntegral (LA.size y))
  return [ BenchRow "haskell" "regression" name ms r2 rmse "fitLM (OLS)" ]

-- ---------------------------------------------------------------------------
-- GLM (Logistic / Poisson, IRLS)
-- ---------------------------------------------------------------------------

benchLogistic :: FilePath -> String -> IO [BenchRow]
benchLogistic = benchGLM Binomial Logit "fitGLM Binomial/Logit"

benchPoisson :: FilePath -> String -> IO [BenchRow]
benchPoisson = benchGLM Poisson Log "fitGLM Poisson/Log"

benchGLM
  :: Family -> LinkFn -> String -> FilePath -> String -> IO [BenchRow]
benchGLM fam link extra path name = do
  (x, y) <- readCsvXY path
  let xWith1 = LA.fromBlocks [[ LA.konst 1 (LA.rows x, 1), x ]]
  (ms, fr) <- timeitTastyIO forceFR
                (\i -> return $! fitGLMFullPhantom i fam link xWith1 y)
  let yhat = LA.flatten (fitted fr LA.¿ [0])
      r2   = computeR2 y yhat
      rmse = sqrt (LA.sumElements ((y - yhat) ** 2) / fromIntegral (LA.size y))
  return [ BenchRow "haskell" "regression" name ms r2 rmse extra ]

-- ---------------------------------------------------------------------------
-- LME (Gaussian, exact EM)
-- ---------------------------------------------------------------------------

benchLME :: FilePath -> String -> IO [BenchRow]
benchLME path name = do
  (x, gIdxs, y) <- readCsvXYG path
  let xWith1 = LA.fromBlocks [[ LA.konst 1 (LA.rows x, 1), x ]]
      uniq   = uniqueInts (V.toList gIdxs)
      labels = V.fromList (map (T.pack . ('g' :) . show) uniq)
      sizes  = V.fromList
        [ length (filter (== g) (V.toList gIdxs)) | g <- uniq ]
  (ms, fit) <- timeitTastyIO (\f -> LA.sumElements (coefficients (GLMM.glmmFixed f)))
                (\i -> return $! fitLMEPhantom i xWith1 y gIdxs labels sizes)
  let yhat = LA.flatten (fitted (GLMM.glmmFixed fit) LA.¿ [0])
      r2   = computeR2 y yhat
  return
    [ BenchRow "haskell" "regression" name ms r2 (GLMM.glmmICC fit)
        "fitLME exact EM" ]
  where
    uniqueInts = foldr (\a acc -> if a `elem` acc then acc else a : acc) []

-- ---------------------------------------------------------------------------
-- Ridge / Lasso / ElasticNet
-- ---------------------------------------------------------------------------

benchRidge :: FilePath -> String -> Double -> IO [BenchRow]
benchRidge = benchPenalty (\lam -> Reg.L2 lam)
                          (\lam -> "fitRidge lambda=" ++ show lam)

benchLasso :: FilePath -> String -> Double -> IO [BenchRow]
benchLasso = benchPenalty (\lam -> Reg.L1 lam)
                          (\lam -> "fitLasso lambda=" ++ show lam ++ " (CD)")

benchEN :: FilePath -> String -> Double -> Double -> IO [BenchRow]
benchEN path name lam1 lam2 = do
  (x, y) <- readCsvXY path
  (ms, fr) <- timeitTastyIO forceReg
                (\i -> return $! fitRegPhantom i (Reg.ElasticNet lam1 lam2) x y)
  let yhat = Reg.predictRegularized fr x
      r2   = computeR2 y yhat
      rmse = sqrt (LA.sumElements ((y - yhat) ** 2) / fromIntegral (LA.size y))
  return [ BenchRow "haskell" "regression" name ms r2 rmse
                     ("fitElasticNet lam1=" ++ show lam1
                      ++ " lam2=" ++ show lam2) ]

benchPenalty
  :: (Double -> Reg.Penalty)
  -> (Double -> String)
  -> FilePath -> String -> Double -> IO [BenchRow]
benchPenalty mkPen mkExtra path name lam = do
  (x, y) <- readCsvXY path
  (ms, fr) <- timeitTastyIO forceReg
                (\i -> return $! fitRegPhantom i (mkPen lam) x y)
  let yhat = Reg.predictRegularized fr x
      r2   = computeR2 y yhat
      rmse = sqrt (LA.sumElements ((y - yhat) ** 2) / fromIntegral (LA.size y))
  return [ BenchRow "haskell" "regression" name ms r2 rmse (mkExtra lam) ]

-- ---------------------------------------------------------------------------

forceFR :: FitResult -> Double
forceFR fr = LA.sumElements (coefficients fr)
           + LA.sumElements (residuals fr)

forceReg :: Reg.RegFit -> Double
forceReg fr = LA.sumElements (Reg.rfBeta fr)
            + LA.sumElements (Reg.rfYHat fr)

computeR2 :: LA.Vector Double -> LA.Vector Double -> Double
computeR2 y yhat =
  let mu  = LA.sumElements y / fromIntegral (LA.size y)
      sst = LA.sumElements ((y - LA.konst mu (LA.size y)) ** 2)
      sse = LA.sumElements ((y - yhat) ** 2)
  in if sst == 0 then 0 else 1 - sse / sst
